### DISCUSSION TITLE
How to stop world.run before end by command/message/callback ?

### DISCUSSION THREAD

**Discussion by aguag** [OFFIS-mosaik/mosaik#6] [Q&A] [answered]
Hi, We are developing a mosaik application that can run a simulation in real-time (with 1-second step_size) basically scheduled in an endless loop ( world.run(until=END,rt_factor=1.) with END parameter very, very big). This implies that in case we need to change any parameters of the simulators, we have to stop by code the world.run,  e.g. by a command/method, and then restart it.  I have not found any example of doing it. I understand that there is the possibility to use instead of world.run other sim manager methods, but in this case, we need an example to clarify the usage. Can anybody help us? Which is the best practice in this case? Any suggestion is appreciated!

**Comment by eike** [comment, marked as answer]
Unfortunately, this is not supported by mosaik natively, yet. I also do not see a good way of doing this via internal functions. (You could just have your simulator close its connection which will be interpreted as an error by mosaik, which will in turn end the simulation. But this will not be very graceful and you need to be careful that all needed information is processed by then.)

If you leave a more detailed description of your usecase, we might consider expanding mosaik in the necessary way. In particular, it would be interesting to know where the decision to restart is made.

**Reply by aguag** [reply]
Thank you for your answer. Perhaps the ability to stop world.run from running could be useful to other users as well. Our application Our application intends to use Mosaik as a scheduler of a process that is always running. It is a Digital Twin of a battery which, among the various operating modes, the two main ones are:

1) Learning mode, the model iteratively reads measurements from the real battery and tunes its parameters with a machine learning algorithm in order to correct the static and dynamic characteristics of the model that change as the battery itself ages. In this mode the parameters of the model will be calculated which will therefore better copy the behavior of the battery.

2) Simulation mode. In this mode the battery model is used as a simulator whose model will be more reliable thanks to the learning mode activity. In this mode you could operate in real time. In this way, for example, the user could change the load and check the voltage trend at the battery terminals. Among other things, the simulator could be used as an estimator of the battery state given that the model can also identify internal battery measurements that cannot be measured directly on the real battery.

The need to be able to interrupt the run is clear: for example, once the learning activity is concluded it is necessary to stop world.run. reload the new parameters on the battery model, re-initialize and then re-launch the battery class constructor.
Finally, a further useful feature to introduce into Mosaik (maybe there is???), more complex to implement and perhaps too specialized, would be the hot-stop of a single simulator, which therefore leaves the other simulators in run and the subsequent hot-run of only one simulator

**Reply by eike** [reply]
Ah, I see. My first approach to the "replace one simulator by a twin" challenge would have been to do that within the mosaik simulator (i.e. have one simulator that uses the real battery or the twin based on its configuration, which it could change mid-run). That said, it seems like something that might come up occasionally, so a dedicated mosaik feature could be worthwhile.

To understand your use-case a bit better: Where would the decision to switch from learning mode to simulation mode take place? In a simulator or in an (maybe even human?) "observer" of the simulation? The design of the stopping mechanism would need to be vastly different in these two cases.

**Reply by aguag** [reply]
The switch is a decision of the human that operates with the HMI of the digital twin. A Redis server is used to share measures, states and commands between HMI, Mosaik simulators, and the real battery.
The application will be totally containarized with Docker. In this regard, is it possible to have more instances of Mosaik running at the same time? (different port number?). There are "support simulators", task in the Digital Twin Mosaik scenario such as the one that takes care of saving historical data on a influxdb server, that don't need to be stopped, indeed they must never be stopped. And in general there are task of the application that in some cases may or have to be stopped and other that have to be always running. 
We have not already taken a final decision but in case we choose a more extensive use of Mosaik also for these support services which have a different execution policy compared to the modeling ones, we may think of having two instances of Mosaik, one that collects the stoppable tasks, the other the not-stoppable or other-execution-policy tasks.

**Reply by eike** [reply]
It is possible to have multiple instances of mosaik running at the same time by changing the port via the `mosaik_config` parameter to `mosaik.World`'s constructor, see here: https://mosaik.readthedocs.io/en/latest/scenario-definition.html#the-setup
